pkg/config/expression.go

```diff
@@ -64,47 +67,216 @@ func SimpleVarToReference(s string) (Reference, error) {
+func IsRawHclLiteral(v cty.Value) (string, bool) {
+	if v.Type() != cty.String {
+		return "", false
+	}
+	s := v.AsString()
+	if len(s) < 4 || s[:2] != "((" || s[len(s)-2:] != "))" {
+		return "", false
+	}
+	return s[2 : len(s)-2], true
+}
+
+// HclExpression is a representation of HCL literals in Blueprint
+type HclExpression struct {
+	e  hclsyntax.Expression
+	s  string
+	rs []Reference
+}
+
+// ParseExpression returns HclExpression
+func ParseExpression(s string) (HclExpression, error) {
+	e, diag := hclsyntax.ParseExpression([]byte(s), "", hcl.Pos{})
+	if diag.HasErrors() {
+		return HclExpression{}, diag
+	}
+	ts := e.Variables()
+	rs := make([]Reference, len(ts))
+	for i, t := range ts {
+		var err error
+		if rs[i], err = TraversalToReference(t); err != nil {
+			return HclExpression{}, err
+		}
+	}
+	return HclExpression{e: e, s: s, rs: rs}, nil
+}
+
+// MustParseExpression is "errorless" version of ParseExpression
+// NOTE: only use it if passed expression is guaranteed to be correct
+func MustParseExpression(s string) HclExpression {
+	if exp, err := ParseExpression(s); err != nil {
+		panic(fmt.Errorf("error while parsing %#v: %w", s, err))
+	} else {
+		return exp
+	}
+}
+
+// Tokenize returns Tokens to be used for marshalling HCL
+func (e HclExpression) Tokenize() hclwrite.Tokens {
+	return hclwrite.TokensForIdentifier(e.s)
+}
+
+// References return Reference for all variables used in the expression
+func (e HclExpression) References() []Reference {
+	c := make([]Reference, len(e.rs))
+	for i, r := range e.rs {
+		c[i] = r
+	}
+	return c
 }
 
-// DoNotAllowVariablesTranslator does not do any translation, it raises an error if any variables are met
-type DoNotAllowVariablesTranslator struct {
-	VariableTranslator
+// AsRawValue returns a cty.Value, that is rendered as
+// HCL literal in Blueprint syntax. Returned value isn't functional,
+// as it doesn't reference HclExpression.
+// This method should only be used for marshaling Blueprint YAML.
+func (e HclExpression) AsRawValue() cty.Value {
+	return cty.StringVal("((" + e.s + "))")
 }
 
-// TranslateSimpleToHcl raises an error
-func (t DoNotAllowVariablesTranslator) TranslateSimpleToHcl(s string) (string, error) {
-	return "", fmt.Errorf("variables aren't allowed here, got %#v", s)
+// To associate cty.Value with HclExpression we use cty.Value.Mark
+// See: https://pkg.go.dev/github.com/zclconf/go-cty/cty#Value.Mark
+// "Marks" should be of hashable type, sadly HclExpression isn't one.
+// We work it around by using `hclExpressionKey` - unique identifier
+// of HclExpression and global map of used expressions `globalHclExpressions`.
+// There are two guarantees to hold:
+// * `ex1.key() == ex2.key()` => `ex1` and `ex2` are identical.
+// It's achieved by using HCL literal as a key.
+// * Every "mark" is in agreement with `globalHclExpressions`.
+// Achieved by declaring `HclExpression.AsValue()` as the ONLY way to produce "marks".
+type hclExpressionKey struct {
```

Comment:
this is a useful comment. Should I be concerned that two expressions with identical HCL literals but different Marks may collide?

e.g. one set by the user explicitly with `$(vars.value)` and another with `use` keyword?

Reply:
Marks are attributes of `cty.Value` and not has no relation to `HclExpression`.
So one CAN NOT answer question "Is this expression result of `use` keyword", instead one should ask "Is this value in module settings is  result of `use` keyword"